Document the new OpenLDAP configuration feature for Okta auth providers (#599)
rancher/rancher-docs: docs/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-okta-saml.md

- Users and groups aren't validated when you assign permissions to them in Rancher.
- When adding users, the exact user IDs (i.e. `UID Field`) must be entered correctly. As you type the user ID, there will be no search for other  user IDs that may match.
- When adding groups, you must select the group from the drop-down that is next to the text box. Rancher assumes that any input from the text box is a user.
- The group drop-down shows only the groups that you are a member of. You will not be able to add groups that you are not a member of.

:::

## Okta with OpenLDAP search

You can add an OpenLDAP backend to assist with user and group search. Rancher will display additional users and groups from the OpenLDAP service. This allows assigning permissions to groups that the logged-in user is not already a member of.

### OpenLDAP Prerequisites

If you use Okta as your IdP, you can [set up an LDAP interface](https://help.okta.com/en-us/Content/Topics/Directory/LDAP-interface-main.htm) for Rancher to use. You can also configure an external OpenLDAP server.

> You must configure Rancher with a LDAP bind account (aka service account) so that you can search and retrieve LDAP entries for users and groups that should have access. Don't use an administrator account or personal account as an LDAP bind account. Create a dedicated account in OpenLDAP, with read-only access to users and groups under the configured searchbase.

Reviewer: ```suggestion
You must configure Rancher with a LDAP bind account (aka service account) so that you can search and retrieve LDAP entries for users and groups that should have access. Don't use an administrator account or personal account as an LDAP bind account. [Create](https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-add-users.htm) a dedicated account in OpenLDAP, with read-only access to users and groups under the configured searchbase.
```

Link to instructions on how to create/add a new user in Okta

> You must configure Rancher with a LDAP bind account (aka service account) so that you can search and retrieve LDAP entries for users and groups that should have access. Don't use an administrator account or personal account as an LDAP bind account. [Create](https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-add-users.htm) a dedicated account in OpenLDAP, with read-only access to users and groups under the configured searchbase.

Reviewer 2: Can you add a warning somewhere that this feature will allow all users to see groups accessible by the service account?

> **Using TLS?**
>

> > If the certificate used by the OpenLDAP server is self-signed or not from a recognized certificate authority, Rancher needs the CA certificate (concatenated with any intermediate certificates) in PEM format. You will have to provide this certificate during the configuration so that Rancher is able to validate the certificate chain.

Reviewer: ```suggestion
> If the certificate used by the OpenLDAP server is self-signed or from an unrecognized certificate authority, Rancher needs the CA certificate (concatenated with any intermediate certificates) in PEM format. Provide this certificate during the configuration so that Rancher can validate the certificate chain.
```

:::note SAML Provider Caveats:

If you configure Okta without OpenLDAP, the following caveats apply due to the fact that SAML Protocol does not support search or lookup for users or groups.

> - There is no validation on users or groups when assigning permissions to them in Rancher.

Author: The phrase "use OpenLDAP as your Okta backend" ends up being extra confusing, but mostly because Okta itself is more flexible. It can either use its own directory layered with an LDAP _interface_, or it can be set up as a wrapper around some backing identity provider, which may either _be_ OpenLDAP, or provide its own separate LDAP interface. Rancher doesn't especially care how the various IdP bits link together, it just needs SAML details and LDAP details that map to the same users.

Since all of these varied flows are supported, we should try to avoid suggesting that a specific backend configuration is required.

> - There is no validation on users or groups when assigning permissions to them in Rancher.

Reviewer: Ok, I wasn't aware of the distinction. You can ignore the suggestion.

- When adding users, the exact user IDs (i.e. `UID Field`) must be entered correctly. As you type the user ID, there will be no search for other  user IDs that may match.
- When adding groups, you must select the group from the drop-down that is next to the text box. Rancher assumes that any input from the text box is a user.
- The group drop-down shows only the groups that you are a member of. You will not be able to add groups that you are not a member of.

:::

> To enable group and user lookup when assigning permissions in Rancher, you will need to configure a back end for the SAML provider that supports groups, such as OpenLDAP.

Author: Oh, yeah when I copied this from the shibboleth document I wasn't aware that Rancher is already doing local group search. This is written weirdly and redundant with the above blurb, so I'll just remove the line entirely.

> To enable searching for groups when assigning permissions in Rancher, you will need to configure a back end for the SAML provider that supports groups, such as OpenLDAP.

Reviewer: ```suggestion
To enable group and user lookup when assigning permissions in Rancher, you need to configure a backend for the SAML provider that supports groups, such as OpenLDAP.
```

## Setting up OpenLDAP in Rancher

If you also configure OpenLDAP as the back end to Okta, it will return a SAML assertion to Rancher with user attributes that include groups. Then authenticated users will be able to access resources in Rancher that their groups have permissions for.

### OpenLDAP Prerequisites

> If you use an Okta Directory as your IdP, you can setup an LDAP Interface for Rancher to use. See the [Okta Documentation](https://help.okta.com/en-us/Content/Topics/Directory/LDAP-interface-main.htm). You can also configure an external OpenLDAP server.

Reviewer: ```suggestion
If you use Okta as your IdP, you can [set up an LDAP interface](https://help.okta.com/en-us/Content/Topics/Directory/LDAP-interface-main.htm) for Rancher to use. You can also configure an external OpenLDAP server.
```

> Rancher must be configured with a LDAP bind account (aka service account) to search and retrieve LDAP entries pertaining to users and groups that should have access. It is recommended to not use an administrator account or personal account for this purpose and instead create a dedicated account in OpenLDAP with read-only access to users and groups under the configured search base (see below).

Author: OpenLDAP doesn't necessarily have the _concept_ of users. That's an optional schema; under the hood it's basically a tree-like structure and exceptionally malleable. Actual instructions for user creation will depend on the user's true backend. For example, Okta can provide an LDAP Interface and, if using this, the service account is just a regular user added to the Okda Directory.

> Rancher must be configured with a LDAP bind account (aka service account) to search and retrieve LDAP entries pertaining to users and groups that should have access. It is recommended to not use an administrator account or personal account for this purpose and instead create a dedicated account in OpenLDAP with read-only access to users and groups under the configured search base (see below).

Reviewer: Ah, that's significantly more complicated than what I had in mind (simply linking the user to a pre-existing external page with instructions on how to create a dedicated account). But since we're specifically discussing Okta here, maybe we can link to this? https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-add-users.htm

> **Using TLS?**
>
> If the certificate used by the OpenLDAP server is self-signed or not from a recognized certificate authority, make sure have at hand the CA certificate (concatenated with any intermediate certificates) in PEM format. You will have to paste in this certificate during the configuration so that Rancher is able to validate the certificate chain.

### Configure OpenLDAP in Rancher

Configure the settings for the OpenLDAP server, groups and users. For help filling out each field, refer to the [configuration reference.](/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/configure-openldap/openldap-config-reference.md) Note that nested group membership is not available.

> Before you proceed with the configuration, please familiarise yourself with the concepts of [External Authentication Configuration and Principal Users](/pages-for-subheaders/authentication-config.md#external-authentication-configuration-and-principal-users).

> 1. Log into the Rancher UI using the initial local `admin` account.

Reviewer: I think here we should let users know they can use a local account with admin privileges, as specifying the *initial* local admin might cause confusion (especially if the user doesn't have access to that account for some reason)
